devito/targets/basic.py

```diff
@@ -0,0 +1,153 @@
+        """The rewrite passes."""
+        return
+
+
+class TargetsMap(OrderedDict, metaclass=Singleton):
+
+    """
+    A special mapper for Targets:
+
+        (platform, mode, language) -> target
+
+    where:
+
+        * `platform` is an object of type Platform, that is the architecture
+          the generated code should be specializer for.
+        * `mode` is the optimization level (e.g., `advanced`)
+        * `language` is the generated code language (default is C+OpenMP+MPI,
+          but in the future it could also be OpenACC or CUDA.
+        * `target` is an object of type Target.
+    """
+
+    _modes = ('noop', 'advanced')
+    _languages = ('C',)
+    _registry = _modes + tuple(product(_modes, _languages))
+
+    def add(self, target, platform, mode, language='C'):
+        assert issubclass(target, Target)
+        assert issubclass(platform, Platform)
+        assert mode in TargetsMap._modes or mode == 'custom'
+
+        self[(platform, mode, language)] = target
+
+    def fetch(self, platform, mode, language='C'):
+        # Retrieve the most specialized Target
+        for cls in platform.__class__.mro():
+            for (p, m, l), target in self.items():
+                if issubclass(p, cls) and m == mode and l == language:
+                    return target
+        raise KeyError("Couldn't find a Target for `%s`" % str((p, m, l)))
+
+
+targets = TargetsMap()
+"""To be populated by the individual backends."""
+
+targets_registry = TargetsMap._registry
+
+
+def iet_lower(iet, mode='advanced', options=None):
+    """
+    Specialize an Iteration/Expression tree (IET) for a certain Target. This
+    consists of:
+
+        * Loop-level transformations to generate Target-optimized code;
+        * Shared-memory and distributed-memory parallelism;
+        * Symbol and data management.
+
+    Parameters
+    ----------
+    iet : Node
+        The IET to be lowered.
+    mode : str, optional
+        The DLE transformation mode.
+        - ``noop``: Do nothing.
+        - ``advanced``: flush denormals, vectorization, loop blocking, OpenMP-related
+                        optimizations (collapsing, nested parallelism, ...), MPI-related
+                        optimizations (aggregation of communications, reshuffling of
+                        communications, ...).
+    options : dict, optional
+        - ``openmp``: Enable/disable OpenMP. Defaults to `configuration['openmp']`.
+        - ``mpi``: Enable/disable MPI. Defaults to `configuration['mpi']`.
+        - ``blockinner``: Enable/disable blocking of innermost loops. By default,
+                          this is disabled to maximize SIMD vectorization. Pass True
+                          to override this heuristic.
+        - ``blocklevels``: Levels of blocking for hierarchical tiling (blocks,
+                           sub-blocks, sub-sub-blocks, ...). Different Platforms have
+                           different default values.
```

Comment:
Add : ...Defaults to...

Reply:
This docstring is changing in one of the follow up PRs, so let's just wait for it I'd say